File: src/SnoDb/SnoDatabase.vb
Imports System.IO
Imports System.IO.Compression
Imports System.Collections.Concurrent

Public Class SnoDatabase

    Public ReadOnly Property Name As String
    Public ReadOnly Property DatabaseDirectory As String
    Private ReadOnly Property ArchivePath As String
    Private ReadOnly Property Collections As ConcurrentDictionary(Of String, ISnoCollection) = New ConcurrentDictionary(Of String, ISnoCollection)()

    Public Sub New(name As String)
        Me.New(name, Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.CommonApplicationData), "SnoDb\Databases\"))
    End Sub

    Public Sub New(name As String, databaseDirectory As String)
        Me.Name = name
        Me.DatabaseDirectory = databaseDirectory

        Directory.CreateDirectory(Me.DatabaseDirectory)
        Me.ArchivePath = Path.Combine(Me.DatabaseDirectory, $"{name}.zip")
    End Sub

    Public Function RegisterCollection(Of T, TId)(idSelector As Func(Of T, TId)) As ISnoCollection(Of T)
        Dim collectionName = Me.GetCollectionName(Of T)()
        Dim collection = New SnoCollection(Of T, TId)(Me, collectionName, idSelector)

        Me.Collections.TryAdd(collectionName, collection)

        Return collection
    End Function

    Public Function GetCollection(Of T)() As ISnoCollection(Of T)
        Dim collectionName = Me.GetCollectionName(Of T)()
        Dim collection = Me.Collections(collectionName)

        Return DirectCast(collection, ISnoCollection(Of T))
    End Function

    Public Async Function LoadAsync() As Task
        If Not File.Exists(Me.ArchivePath) Then
            Return
        End If

        Using archive = ZipFile.Open(Me.ArchivePath, ZipArchiveMode.Read)
            Dim entriesByCollection = (From entry In archive.Entries
                                       Let separatorIndex = entry.FullName.IndexOf("/"c)
                                       Let collectionName = entry.FullName.Substring(0, separatorIndex)
                                       Group entry By collectionName Into Group
                                       Select New With {.CollectionName = collectionName, .Entries = Group})

            For Each collectionEntries In entriesByCollection
                Dim currentCollection As ISnoCollection = Nothing
                If Me.Collections.TryGetValue(collectionEntries.CollectionName, currentCollection) Then
                    Await currentCollection.LoadAsync(collectionEntries.Entries)
                End If
            Next
        End Using
    End Function

    Public Async Function SaveAsync() As Task
        Using archive = ZipFile.Open(Me.ArchivePath, ZipArchiveMode.Update)
            For Each collection In Me.Collections.Values
                Await collection.SaveAsync(archive)
            Next
        End Using
    End Function

    Private Function GetCollectionName(Of T)() As String
        Return GetType(T).Name
    End Function

End Class

File: src/SnoDb/SnoDatabase.cs
using System;
using System.Linq;
using System.Collections.Concurrent;
using System.IO;
using Ionic.Zip;

namespace Symmex.SnoDb
{
    public class SnoDatabase : IDisposable
    {
        public string Name { get; }
        public string DatabaseDirectory { get; }
        private ZipFile Archive { get; }
        private ConcurrentDictionary<string, ISnoCollection> Collections { get; } = new ConcurrentDictionary<string, ISnoCollection>();
        private bool IsLoaded { get; set; }

        public SnoDatabase(string name)
            : this(name, SnoDbConfig.DefaultDatabaseDirectory)
        {
        }

        public SnoDatabase(string name, string databaseDirectory)
        {
            Name = name;
            DatabaseDirectory = databaseDirectory;

            Directory.CreateDirectory(DatabaseDirectory);

            var archivePath = Path.Combine(DatabaseDirectory, $"{Name}.zip");
            Archive = new ZipFile(archivePath);
        }

        private string GetCollectionName<T>()
        {
            return typeof(T).Name;
        }

        public ISnoCollection<T> RegisterCollection<T, TId>(Func<T, TId> idSelector)
        {
            EnsureLoaded();

            var collectionName = GetCollectionName<T>();
            var collection = new SnoCollection<T, TId>(this, collectionName, idSelector);
            Collections.TryAdd(collectionName, collection);

            return collection;
        }

        public ISnoCollection<T> GetCollection<T>()
        {
            EnsureLoaded();

            var collectionName = GetCollectionName<T>();
            var collection = Collections[collectionName];

            return (ISnoCollection<T>)collection;
        }

        private void EnsureLoaded()
        {
            if(!IsLoaded)
                Load();
        }

        public void Load()
        {
            var entriesByCollection = (from entry in Archive.Entries
                                       let separatorIndex = entry.FileName.IndexOf('/')
                                       let collectionName = entry.FileName.Substring(0, separatorIndex)
                                       group entry by collectionName into g
                                       select new { CollectionName = g.Key, Entries = g.ToList() });

            foreach (var collectionEntries in entriesByCollection)
            {
                ISnoCollection currentCollection;
                if (Collections.TryGetValue(collectionEntries.CollectionName, out currentCollection))
                {
                    currentCollection.Load(collectionEntries.Entries);
                }
            }

            IsLoaded = true;
        }

        public void Save()
        {
            foreach (var collection in Collections.Values)
            {
                collection.Save(Archive);
            }

            Archive.Save();
        }

        #region IDisposable Support
        private bool disposedValue = false; // To detect redundant calls

        protected virtual void Dispose(bool disposing)
        {
            if (!disposedValue)
            {
                if (disposing)
                {
                    Save();
                    Archive.Dispose();
                }

                disposedValue = true;
            }
        }

        public void Dispose()
        {
            Dispose(true);
        }
        #endregion
    }
}
